### ISSUE TITLE
Should we actually run fixtures concurrently?

### ISSUE THREAD

**Issue by njsmith** [python-trio/pytest-trio#57]
Upsides:

* Can make test setup/teardown faster
* Since we have to run fixtures in independent tasks anyway (see #55), running them concurrently was easier than running them sequentially.

Counterarguments:

* Might increase the chance of flaky tests, race conditions, etc.
  * In particular, race conditions involving `ContextVars` (see https://github.com/python-trio/pytest-trio/pull/50#issuecomment-414648141) – currently fixtures all share a `contextvars.Context`, so if multiple fixtures are mutating the same `ContextVar`, the results will be random. (Of course, doing this is inherently nonsensical. But nonsensical-and-consistent is still better than nonsensical-and-random.)
* Might make failing tests harder to debug
* Is there really that much concurrency to extract from fixture setup? How many people use multiple independent fixtures that need to block for significant amounts of time?
* Running them sequentially isn't actually that hard, if we let pytest pick the order for us (see https://github.com/python-trio/pytest-trio/pull/50#issuecomment-414652134)

Leaving this issue open for now for discussion, and to collect any positive/negative experiences people want to report.

**Comment by njsmith** [comment]
#120 is a genuine bug that is partly due to the complexity of managing concurrent fixture startup -- specifically, if one fixture setup crashes, we stop creating new ones, but we don't do anything to cancel any concurrent fixtures. And fixing it is made more difficult by the complexity of our fixture setup code. If we switched to serial setup, it might make this more maintainable.

**Comment by gc-ss** [comment]
> If we switched to serial setup, it might make this more maintainable.

I don't think we had to take that route. The fix was ensuring the cancel scopes got passed through properly instead?

**Comment by njsmith** [comment]
Not so much passing them through properly, but keeping track of which fixtures were currently being set up, and cancelling all of them instead of just the test.
